src/coreclr/src/vm/nativeimage.h

```diff
@@ -0,0 +1,108 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _NATIVEIMAGE_H
+#define _NATIVEIMAGE_H
+
+#include "readytoruninfo.h"
+
+// This structure is used in NativeImage to map simple names of component assemblies
+// to their indices within the component assembly header table.
+struct AssemblyNameIndex
+{
+    LPCUTF8 Name;
+    int32_t Index;
+    
+    AssemblyNameIndex() : Name(NULL), Index(-1) {}
+    AssemblyNameIndex(LPCUTF8 name, int32_t index) : Name(name), Index(index) {}
+    
+    static AssemblyNameIndex GetNull() { return AssemblyNameIndex(); }
+    bool IsNull() const { return Index < 0; }
+};
+
+class AssemblyNameIndexHashTraits : public NoRemoveSHashTraits< DefaultSHashTraits<AssemblyNameIndex> >
+{
+public:
+    // Similar to BaseAssemblySpec::CompareStrings, we're using temporary SStrings that throw
+    // for case-insensitive UTF8 assembly name comparisons.
+    static const bool s_NoThrow = false;
+
+    typedef LPCUTF8 key_t;
+
+    static AssemblyNameIndex Null() { return AssemblyNameIndex::GetNull(); }
+    static bool IsNull(const AssemblyNameIndex& e) { return e.IsNull(); }
+
+    static LPCUTF8 GetKey(const AssemblyNameIndex& assemblyNameIndex) { return assemblyNameIndex.Name; }
+    static BOOL Equals(LPCUTF8 a, LPCUTF8 b);
+    static count_t Hash(LPCUTF8 a);
+};
+
+class AssemblyLoadContext;
+class ReadyToRunInfo;
+class PEFile;
+class PEImage;
+
+// This class represents a  ReadyToRun image with native OS-specific envelope. As of today,
+// this file format is used as the compiled native code cache in composite R2R Crossgen2
+// build mode. Moving forward we plan to add support for OS-specific native executables
+// (ELF on Linux, MachO on OSX).
+//
+// The native image is identified by a well-known public export 'RTR_HEADER' pointing to the
+// master READYTORUN_HEADER structure for the entire file. For composite R2R executables
+// built by crossgenning a larger number of input MSIL assemblies the READYTORUN_HEADER
+// contains a section named ComponentAssemblies that points to READYTORUN_CORE_HEADER
+// structures representing the individual component assemblies and per-assembly sections.
+class NativeImage
+{
+private:
+    // Points to the OwnerCompositeExecutable section content within the component MSIL module
+    LPCUTF8 m_utf8SimpleName;
```

Comment:
Is this a simple name (ie binary name without any prefix/suffix) or is this actually a filename? If I am reading the code right, it should be a filename.

Reply:
You're right, it is supposed to include the extension. Renamed to m_utf8FileName in 16th commit.

Reply:
I would just call it `m_pFileName`. I do not think we have `m_utf8XXX` anywhere else.